File: frontend/src/api/campus.ts
/**
 * Campus API - 校区和教室相关接口
 */
import request from './index'
import type { ResponseModel } from '@/types'

// 校区接口
export interface Campus {
  id: number
  name: string
  address?: string
  phone?: string
  is_active: boolean
  created_at: string
  updated_at: string
  classroom_count?: number
}

export interface CampusCreate {
  name: string
  address?: string
  phone?: string
  is_active?: boolean
}

export interface CampusUpdate {
  name?: string
  address?: string
  phone?: string
  is_active?: boolean
}

// 教室接口
export interface Classroom {
  id: number
  campus_id: number
  name: string
  capacity?: number
  description?: string
  equipment?: Record<string, string>  // 设备配置：key-value形式
  is_active: boolean
  created_time: string
  updated_time: string
  campus?: Campus
}

export interface ClassroomCreate {
  campus_id: number
  name: string
  capacity?: number
  description?: string
  equipment?: Record<string, string>  // 设备配置：key-value形式
  is_active?: boolean
}

export interface ClassroomUpdate {
  campus_id?: number
  name?: string
  capacity?: number
  description?: string
  equipment?: Record<string, string>  // 设备配置：key-value形式
  is_active?: boolean
}

// 分页响应
export interface PaginatedResponse<T> {
  items: T[]
  total: number
  page: number
  page_size: number
  pages: number
}

// 校区 API - 后端用success_response包装数组会变成{items: [...]}格式
export const getCampuses = (params?: {
  include_inactive?: boolean
}) => request.get<ResponseModel<{items: Campus[]}>>('/campuses', { params }).then(res => res.data.data?.items || [])

export const getAllCampuses = (activeOnly = true) =>
  request.get<ResponseModel<{items: Campus[]}>>('/campuses/all', { params: { active_only: activeOnly } }).then(res => res.data.data?.items || [])

export const getCampus = (id: number) =>
  request.get<ResponseModel<Campus>>(`/campuses/${id}`).then(res => res.data.data)

export const createCampus = (data: CampusCreate) =>
  request.post<ResponseModel<Campus>>('/campuses', data).then(res => res.data.data)

export const updateCampus = (id: number, data: CampusUpdate) =>
  request.put<ResponseModel<Campus>>(`/campuses/${id}`, data).then(res => res.data.data)

export const deleteCampus = (id: number) =>
  request.delete(`/campuses/${id}`).then(res => res.data.data)

// 教室 API - 后端用success_response包装数组会变成{items: [...]}格式
export const getClassrooms = (params?: {
  active_only?: boolean
}) => request.get<ResponseModel<{items: Classroom[]}>>('/classrooms', { params }).then(res => res.data.data?.items || [])

export const getAllClassrooms = (campusId?: number, activeOnly = true) =>
  request.get<ResponseModel<{items: Classroom[]}>>('/classrooms/all', {
    params: { campus_id: campusId, active_only: activeOnly }
  }).then(res => res.data.data?.items || [])

export const getClassroom = (id: number) =>
  request.get<ResponseModel<Classroom>>(`/classrooms/${id}`).then(res => res.data.data)

export const createClassroom = (data: ClassroomCreate) =>
  request.post<ResponseModel<Classroom>>('/classrooms', data).then(res => res.data.data)

export const updateClassroom = (id: number, data: ClassroomUpdate) =>
  request.put<ResponseModel<Classroom>>(`/classrooms/${id}`, data).then(res => res.data.data)

export const deleteClassroom = (id: number) =>
  request.delete(`/classrooms/${id}`).then(res => res.data.data)

File: backend/app/api/v1/campus.py
"""
Campus and Classroom API endpoints.
"""
from typing import List

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.api.deps import (
    CurrentUser, AdminUser, CampusRead, CampusEdit, CampusDelete,
    ClassroomRead, ClassroomEdit, ClassroomDelete,
    CampusScopedQuery, DBSession
)
from app.services.campus_service import CampusService
from app.schemas.campus import (
    CampusCreate, CampusUpdate, CampusResponse, CampusWithClassroomsResponse,
    ClassroomCreate, ClassroomUpdate, ClassroomResponse
)
from app.schemas.common import MessageResponse, success_response

router = APIRouter(tags=["校区管理"])
campus_scope = CampusScopedQuery()


def get_service(db: AsyncSession = Depends(get_db)) -> CampusService:
    return CampusService(db)


# ============ Campus Endpoints ============

@router.get("/campuses", summary="获取所有校区")
async def get_campuses(
    include_inactive: bool = False,
    service: CampusService = Depends(get_service),
    _: CurrentUser = None,
):
    """Get all campuses with their classrooms."""
    items = await service.get_all_campuses(include_inactive)
    return success_response([CampusWithClassroomsResponse.model_validate(item).model_dump() for item in items])


@router.get("/campuses/all", summary="获取校区下拉列表")
async def get_all_campuses_dropdown(
    active_only: bool = True,
    service: CampusService = Depends(get_service),
    _: CurrentUser = None,
):
    """Get all campuses for dropdown (no classrooms)."""
    items = await service.get_all_campuses_simple(active_only)
    return success_response([CampusResponse.model_validate(item).model_dump() for item in items])


@router.get("/campuses/{campus_id}", summary="获取校区详情")
async def get_campus(
    campus_id: int,
    service: CampusService = Depends(get_service),
    _: CurrentUser = None,
):
    """Get campus by ID with classrooms."""
    campus = await service.get_campus_by_id(campus_id)
    return success_response(CampusWithClassroomsResponse.model_validate(campus).model_dump())


@router.post("/campuses", summary="创建校区")
async def create_campus(
    data: CampusCreate,
    service: CampusService = Depends(get_service),
    current_user: AdminUser = None,
):
    """Create campus (admin only)."""
    campus = await service.create_campus(data, current_user.username)
    return success_response(CampusResponse.model_validate(campus).model_dump())


@router.put("/campuses/{campus_id}", summary="更新校区")
async def update_campus(
    campus_id: int,
    data: CampusUpdate,
    service: CampusService = Depends(get_service),
    current_user: AdminUser = None,
):
    """Update campus (admin only)."""
    campus = await service.update_campus(campus_id, data, current_user.username)
    return success_response(CampusResponse.model_validate(campus).model_dump())


@router.delete("/campuses/{campus_id}", summary="删除校区")
async def delete_campus(
    campus_id: int,
    service: CampusService = Depends(get_service),
    current_user: AdminUser = None,
):
    """Delete campus (admin only). This will also delete all classrooms."""
    await service.delete_campus(campus_id)
    return success_response({"message": "删除成功"})


@router.get("/campuses/{campus_id}/classrooms", summary="获取校区教室")
async def get_classrooms_by_campus(
    campus_id: int,
    active_only: bool = True,
    service: CampusService = Depends(get_service),
    _: CurrentUser = None,
):
    """Get classrooms by campus ID."""
    items = await service.get_classrooms_by_campus(campus_id, active_only)
    return success_response([ClassroomResponse.model_validate(item).model_dump() for item in items])


@router.post("/campuses/{campus_id}/classrooms", summary="创建教室")
async def create_classroom(
    campus_id: int,
    data: ClassroomCreate,
    service: CampusService = Depends(get_service),
    current_user: AdminUser = None,
):
    """Create classroom (admin only)."""
    classroom = await service.create_classroom(campus_id, data, current_user.username)
    return success_response(ClassroomResponse.model_validate(classroom).model_dump())


# ============ Classroom Endpoints ============

@router.get("/classrooms", summary="获取所有教室")
async def get_all_classrooms(
    active_only: bool = True,
    db: DBSession = None,
    current_user: ClassroomRead = None,
):
    """
    Get all classrooms filtered by user's campus scope.
    超管未选校区时返回全部，其他用户按校区过滤。
    """
    service = CampusService(db)
    campus_id = campus_scope.get_campus_filter(current_user)
    items = await service.get_all_classrooms(active_only, campus_id)
    return success_response([ClassroomResponse.model_validate(item).model_dump() for item in items])


@router.get("/classrooms/all", summary="获取教室下拉列表")
async def get_all_classrooms_dropdown(
    active_only: bool = True,
    db: DBSession = None,
    current_user: ClassroomRead = None,
):
    """
    Get classrooms for dropdown filtered by user's campus scope.
    超管未选校区时返回全部，其他用户按校区过滤。
    """
    service = CampusService(db)
    campus_id = campus_scope.get_campus_filter(current_user)
    items = await service.get_all_classrooms(active_only, campus_id)
    return success_response([ClassroomResponse.model_validate(item).model_dump() for item in items])


@router.put("/classrooms/{classroom_id}", summary="更新教室")
async def update_classroom(
    classroom_id: int,
    data: ClassroomUpdate,
    db: DBSession = None,
    current_user: ClassroomEdit = None,
):
    """Update classroom (requires classroom:edit permission)."""
    service = CampusService(db)
    classroom = await service.update_classroom(classroom_id, data, current_user.username)
    return success_response(ClassroomResponse.model_validate(classroom).model_dump())


@router.delete("/classrooms/{classroom_id}", summary="删除教室")
async def delete_classroom(
    classroom_id: int,
    db: DBSession = None,
    current_user: ClassroomDelete = None,
):
    """Delete classroom (requires classroom:delete permission)."""
    service = CampusService(db)
    await service.delete_classroom(classroom_id)
    return success_response({"message": "删除成功"})
